cluster-autoscaler/simulator/scheduling.go

```diff
@@ -0,0 +1,102 @@
+/*
+Copyright 2022 The Kubernetes Authors.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package simulator
+
+import (
+	"fmt"
+
+	apiv1 "k8s.io/api/core/v1"
+	klog "k8s.io/klog/v2"
+	schedulerframework "k8s.io/kubernetes/pkg/scheduler/framework"
+)
+
+// SchedulingSimulator is a helper object for simulating scheduler behavior.
+// TODO(x13n): Reuse this in filter_out_unschedulable.
+type SchedulingSimulator struct {
```

Comment:
The main value of this is hints, which is not useful in every context where we do scheduling simulations (ex. scale-up binpacking). Maybe rename this to make it more clear? HintCachingSchedulingSimulator? Something more concise would be nice if you have any good ideas :)